Q24ECU/core/include/hal/hal_spi.h

```diff
@@ -0,0 +1,304 @@
+
+// Register Reset
+#define SPI1_REG_RESET()                          \
+    do                                            \
+    {                                             \
+        (RCC->APB2RSTR |= RCC_APB2RSTR_SPI1RST);  \
+        (RCC->APB2RSTR &= ~RCC_APB2RSTR_SPI1RST); \
+    } while (0)
+
+#define SPI2_REG_RESET()                          \
+    do                                            \
+    {                                             \
+        (RCC->APB1RSTR |= RCC_APB1RSTR_SPI2RST);  \
+        (RCC->APB1RSTR &= ~RCC_APB1RSTR_SPI2RST); \
+    } while (0)
+
+#define SPI3_REG_RESET()                          \
+    do                                            \
+    {                                             \
+        (RCC->APB1RSTR |= RCC_APB1RSTR_SPI3RST);  \
+        (RCC->APB1RSTR &= ~RCC_APB1RSTR_SPI3RST); \
+    } while (0)
+
+#define SPI4_REG_RESET()                          \
+    do                                            \
+    {                                             \
+        (RCC->APB2RSTR |= RCC_APB2RSTR_SPI4RST);  \
+        (RCC->APB2RSTR &= ~RCC_APB2RSTR_SPI4RST); \
+    } while (0)
+
+// SPI_DeviceMode
+#define SPI_DEVICE_MODE_SLAVE 0
+#define SPI_DEVICE_MODE_MASTER 1
+// SPI_BusConfig
+#define SPI_BUS_CONFIG_FD 1
+#define SPI_BUS_CONFIG_HD 2
+#define SPI_BUS_CONFIG_SIMPLEX_RXONLY 3
+// SPI_SclkSpeed
+#define SPI_SCLK_SPEED_DIV2 0
+#define SPI_SCLK_SPEED_DIV4 1
+#define SPI_SCLK_SPEED_DIV8 2
+#define SPI_SCLK_SPEED_DIV16 3
+#define SPI_SCLK_SPEED_DIV32 4
+#define SPI_SCLK_SPEED_DIV64 5
+#define SPI_SCLK_SPEED_DIV128 6
+#define SPI_SCLK_SPEED_DIV256 7
+// SPI_DFF
+#define SPI_DFF_8BITS 0
+#define SPI_DFF_16BITS 1
+// SPI_CPOL
+#define SPI_CPOL_LOW 0
+#define SPI_CPOL_HIGH 1
+// SPI_CPHA
+#define SPI_CPHA_LOW 0
+#define SPI_CPHA_HIGH 1
+// SPI_SSM
+#define SPI_SSM_DI 0
+#define SPI_SSM_EN 1
+
+// SPI Application States - used for interrupts mostly
+#define SPI_READY 0
+#define SPI_BUSY_IN_RX 1
+#define SPI_BUSY_IN_TX 2
+
+// Possible SPI application events
+#define SPI_EVENT_TX_CMPLT 1
+#define SPI_EVENT_RX_CMPLT 2
+#define SPI_EVENT_OVR_ERR 3
+#define SPI_EVENT_CRC_ERR 4
+
+// Configuration structure for SPIx peripheral
+typedef struct
+{
+    uint8_t SPI_DeviceMode;
+    uint8_t SPI_BusConfig;
+    uint8_t SPI_SclkSpeed;
+    uint8_t SPI_DFF;  // Data frame format
+    uint8_t SPI_CPOL; // Polarity
+    uint8_t SPI_CPHA; // Phase
+    uint8_t SPI_SSM;  // Slave Management (Software or Hardware)
+} SPI_Config_t;
+
+// Handle structure for SPIx peripheral
+typedef struct
+{
+    SPI_TypeDef *pSPIx; // Holds the base address of the SPI peripheral. Struct overlay for the SPI registers.
+    SPI_Config_t SPIConfig;
+    uint8_t *pTxBuffer;
+    uint8_t *pRxBuffer;
+    uint32_t TxLen;
+    uint32_t RxLen;
+    uint8_t TxState;
+    uint8_t RxState;
+} SPI_Handle_t;
+
+// Peripheral clocks setup
+static inline void HAL_SPI_PeriClockControl(SPI_TypeDef *pSPIx, bool enable)
```

Comment:
I should have specified lowercase `hal`

Reply:
I used lowercase everything but if you want to leave it the way it is just close this and ignore it

Reply:
changed